libbeat/template/processor.go

```diff
@@ -0,0 +1,259 @@
+package template
+
+import (
+	"errors"
+
+	"github.com/elastic/beats/libbeat/common"
+)
+
+type Processor struct {
+	EsVersion common.Version
+}
+
+var (
+	defaultScalingFactor = 1000
+)
+
+// This includes all entries without special handling for different versions.
+// Currently this is:
+// long, geo_point, date, short, byte, float, double, boolean
+func (p *Processor) process(fields common.Fields, path string, output common.MapStr) error {
+	for _, field := range fields {
+
+		field.Path = path
+		var mapping common.MapStr
+
+		switch field.Type {
+		case "ip":
+			mapping = p.ip(&field)
+		case "scaled_float":
+			mapping = p.scaledFloat(&field)
+		case "half_float":
+			mapping = p.halfFloat(&field)
+		case "integer":
+			mapping = p.integer(&field)
+		case "text":
+			mapping = p.text(&field)
+		case "", "keyword":
+			mapping = p.keyword(&field)
+		case "object":
+			mapping = p.object(&field)
+		case "array":
+			mapping = p.array(&field)
+		case "group":
+			var newPath string
+			if path == "" {
+				newPath = field.Name
+			} else {
+				newPath = path + "." + field.Name
+			}
+			mapping = common.MapStr{}
+			if field.Dynamic.Value != nil {
+				mapping["dynamic"] = field.Dynamic.Value
+			}
+
+			// Combine properties with previous field definitions (if any)
+			properties := common.MapStr{}
+			key := common.GenerateKey(field.Name) + ".properties"
+			currentProperties, err := output.GetValue(key)
+			if err == nil {
+				var ok bool
+				properties, ok = currentProperties.(common.MapStr)
+				if !ok {
+					// This should never happen
+					return errors.New(key + " is expected to be a MapStr")
+				}
+			}
+
+			if err := p.process(field.Fields, newPath, properties); err != nil {
+				return err
+			}
+			mapping["properties"] = properties
+
+		default:
+			mapping = p.other(&field)
+		}
+
+		if len(mapping) > 0 {
+			output.Put(common.GenerateKey(field.Name), mapping)
+		}
+	}
+	return nil
+}
+
+func (p *Processor) other(f *common.Field) common.MapStr {
```

Comment:
I personally preferred that these were methods of Field as all the parts in this methods are applied to on the passed field anyway. What is the reason you changed this logic? Also this increases the number of params to be passed.

Reply:
I think it makes sense to make the `fields` structure the same and therefore reusable between different processors. 
This is the reason I would not bind processor specific logic to the data structure itself. The logic that is generally applicable for transformation to a ES template and also to a Kibana index pattern could be moved to the `common.fields`. However, I might do this in a seperate step if we encounter processor logic that is generally applicable. 

Another advantage is that this makes it easier to test when you pass a data structure into a method instead having a large interface with lots of methods.

Reply:
Got it. I think we could perhaps somehow compose the overall fields out of `TemplateFields` and `KibanaFields` so the TemplateFields would still keep it's method and not have to be passed each time. But as they overlap this could become trickier.

I think I miss the part on how it simplifies testing.

I suggest we move forward as is right now as functionality will be identical. We can still do refactoring when we have all the changes in.

Reply:
E.g some of the methods use ` p.EsVersion` for decisions on how to transform. I would not consider an ES Version to be a general attribute of a `Field`. 
I fully agree though that we should share logic between `ES` and `Kibana` specific transformations when it overlaps or depends on each other.